nltk/collocations.py

```diff
@@ -122,6 +123,22 @@ def above_score(self, score_fn, min_score):
                 break
 
 
+
+def _build_new_documents(documents, window_size, pad_left=False, pad_right=False, pad_symbol=None):
+        '''
+        Pad the document with the place holder according to the window_size 
+        '''
+        new_documents = iter([])
+        for document in documents:
+            document = iter(document)
+            if pad_left:
+                document = _itertools.chain((pad_symbol,) * (window_size - 1), document)
+            if pad_right:
+                document = _itertools.chain(document, (pad_symbol,) * (window_size - 1))
+            new_documents = _itertools.chain(new_documents, document)
```

Comment:
This creates a left-branching structure such that `n_documents * 2` chain instances are in memory at the same time. Indeed, by iterating through all documents and adding them to the chain, you somewhat defeat the purpose of using iterators. Instead, you can use:
```python
padding = (pad_symbol,) * (window_size - 1)
return _itertools.chain.from_iterable(_itertools.chain(doc, padding) for doc in documents)
```
This will only ever have two `chain` instances not awaiting garbage collection and won't iterate through all documents from the outset.